Split test_c10d.py to test_c10d_common.py, test_c10d_gloo.py, test_c10d_nccl.py

pbelevich opened the pull request:
Test Plan: NA

Differential Revision: D27913170

SciPioneer commented:
Thanks for the refactoring! This will definitely make the test codebase manageable.

On the other hand, I think this PR does a bit more than plain splitting. Therefore, it will be better to add some descriptions in the PR to highlight some changes, e.g., creating some abstract classes like `AbstractCommTest`, `AbstractDistributedDataParallelTest` and `AbstractTimeoutTest`. 
 
 Another suggestion is separating out many fundamental store-related unit tests into a separate file, as these tests and other distributed training tests target different layers.

SciPioneer commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -0,0 +1,2130 @@
+
+class RendezvousEnvTest(TestCase):
+    @retry_on_connect_failures
+    def test_logging_init(self):
```
Why is this test in gloo rather than common?

pbelevich replied:
line 181:
```
        c10d.init_process_group(backend="gloo", init_method="env://")
```

SciPioneer replied:
Better to add `@requires_gloo()` here.

SciPioneer commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -0,0 +1,2130 @@
+            pg.allgather([], [])
+
+        with self.assertRaisesRegex(
+                ValueError, "requires input/output tensor lists to have the same length"
```
The indentation here and many "ValueError" lines are changed. Previously it's aligned with the end of "with". Is this intentional?

SciPioneer commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -0,0 +1,2130 @@
+        train_loop(ddp_withoutload, optimizer_withoutload, 6)
+
+        for p_withload, p_withoutload, p_non_ddp_withload in zip(
+                ddp_withload.parameters(), ddp_withoutload.parameters(), model_withload.parameters()
```
Indentation changed.

pbelevich committed:
Split test_c10d.py to test_c10d_common.py, test_c10d_gloo.py, test_c10d_nccl.py (#56598)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/56598

Test Plan: NA

Reviewed By: SciPioneer

Differential Revision: D27913170

fbshipit-source-id: 4f2c09037b94b37b87545fd2b48ed6502c47433b
